### PR TITLE
WIP: Validator Manager contracts

### PR DESCRIPTION
**Pull request by cam-schultz**
## Why this should be merged
Dev branch for in-progress validator manager contracts. Eligible to merge to `main` once we have release tags of upstreams.

Replaces https://github.com/ava-labs/teleporter/pull/453

## How is this documented

### CODE REVIEW

File: contracts/validator-manager/interfaces/IValidatorManager.sol [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+/**
+ * @notice Event emitted when validator weight is updated.
+ * @param validationID The ID of the validation period
+ * @param nonce The message nonce used to update the validator weight
+ * @param validatorWeight The updated validator weight that is sent to the P-Chain
+ * @param setWeightMessageID The ID of the Warp message that updates the validator's weight on the P-Chain
+ */
+event ValidatorWeightUpdate(
```
Comment (minghinmatthewlam) [inline contracts/validator-manager/interfaces/IValidatorManager.sol:41]: why do we have this event outside of interface definition? Usually it's been included inside the interface

File: contracts/validator-manager/tests/ERC20TokenStakingManagerTests.t.sol [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+} from "../interfaces/IValidatorManager.sol";
+import {PoSValidatorManagerSettings} from "../interfaces/IPoSValidatorManager.sol";
+import {IRewardCalculator} from "../interfaces/IRewardCalculator.sol";
+import {ICMInitializable} from "../../utilities/ICMInitializable.sol";
+import {ExampleERC20} from "@mocks/ExampleERC20.sol";
+import {IERC20} from "@openzeppelin/contracts@5.0.2/token/ERC20/IERC20.sol";
+import {IERC20Mintable} from "../interfaces/IERC20Mintable.sol";
+import {SafeERC20} from "@openzeppelin/contracts@5.0.2/token/ERC20/utils/SafeERC20.sol";
+
+// TODO: Remove this once all unit tests implemented
```
Comment (minghinmatthewlam) [inline contracts/validator-manager/tests/ERC20TokenStakingManagerTests.t.sol:24]: should remove this

File: contracts/validator-manager/tests/NativeTokenStakingManagerTests.t.sol [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    ValidatorRegistrationInput
+} from "../interfaces/IValidatorManager.sol";
+import {PoSValidatorManagerSettings} from "../interfaces/IPoSValidatorManager.sol";
+import {IRewardCalculator} from "../interfaces/IRewardCalculator.sol";
+import {ExampleRewardCalculator} from "../ExampleRewardCalculator.sol";
+import {ICMInitializable} from "../../utilities/ICMInitializable.sol";
+import {INativeMinter} from
+    "@avalabs/subnet-evm-contracts@1.2.0/contracts/interfaces/INativeMinter.sol";
+
+// TODO: Remove this once all unit tests implemented
```
Comment (minghinmatthewlam) [inline contracts/validator-manager/tests/NativeTokenStakingManagerTests.t.sol:22]: remove this

File: contracts/validator-manager/tests/NativeTokenStakingManagerTests.t.sol [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        });
+        // Units tests don't have access to the native minter precompile, so use vm.deal instead.
+        vm.deal(account, account.balance + amount);
+    }
+
+    function _getStakeAssetBalance(address account) internal view override returns (uint256) {
+        return account.balance;
+    }
+}
+// TODO: Remove this once all unit tests implemented
```
Comment (minghinmatthewlam) [inline contracts/validator-manager/tests/NativeTokenStakingManagerTests.t.sol:217]: same here

**Review by geoff-vball — APPROVED** [review]

File: contracts/validator-manager/README.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
+# Validator Manager Contract
+
+> [!CAUTION]
+> The contracts in this directory are still under active development, are unaudited, and should not be used in production.
```
Comment (cam-schultz) [inline contracts/validator-manager/README.md:4]: We can remove this disclaimer once we publish the audit.

**Review by iansuvak — APPROVED** [review]

**Pull request merged**